aten/src/ATen/native/SobolEngineOps.cpp

```diff
@@ -0,0 +1,141 @@
+                                              int64_t dimension, int64_t num_generated) {
+  Tensor wquasi = quasi.clone();
+  std::vector<Tensor> result;
+  AT_CHECK(sobolstate.type().scalarType() == at::ScalarType::Long,
+           "sobolstate needs to be of type ", at::ScalarType::Long);
+  AT_CHECK(wquasi.type().scalarType() == at::ScalarType::Long,
+           "quasi needs to be of type ", at::ScalarType::Long);
+
+  for (int64_t i = 0; i < n; ++i) {
+    int64_t l = rightmost_zero(num_generated);
+    Tensor inter_res = wquasi.__ixor__(sobolstate.select(1, l - 1));
+    result.emplace_back(inter_res);
+    num_generated++;
+  }
+
+  return std::make_tuple(at::native::cat(result, 0).toType(at::kFloat).mul_(RECIPD), wquasi);
+}
+
+/// This is the core function to fast-forward a `SobolEngine` given
+/// its state variables (`sobolstate` and `quasi`). `dimension` can be
+/// inferred from `sobolstate`, but is passed as an argument for the same reasons
+/// specified above.
+Tensor _sobol_engine_ff(const Tensor& quasi, int64_t n, const Tensor& sobolstate,
+                        int64_t dimension, int64_t num_generated) {
+  Tensor wquasi = quasi.clone();
+  AT_CHECK(sobolstate.type().scalarType() == at::ScalarType::Long,
+           "sobolstate needs to be of type ", at::ScalarType::Long);
+  AT_CHECK(wquasi.type().scalarType() == at::ScalarType::Long,
+           "quasi needs to be of type ", at::ScalarType::Long);
+
+  for (int64_t i = 0; i < n; ++i) {
+    int64_t l = rightmost_zero(num_generated);
+    wquasi.__ixor__(sobolstate.select(1, l - 1));
+    num_generated++;
+  }
+  return wquasi;
+}
+
+/// This is an implicit function used for randomizing the state variables of the.
+/// `SobolEngine`. Arguments are a randomized `sobolstate` state variables
+/// and a list of random lower triangular matrices consisting of 0s and 1s. `dimension` is
+/// passed explicitly again.
+Tensor _sobol_engine_scramble(const Tensor& sobolstate, TensorList ltm, int64_t dimension) {
+  Tensor wsobolstate = sobolstate.clone();
+  AT_CHECK(sobolstate.type().scalarType() == at::ScalarType::Long,
+           "sobolstate needs to be of type ", at::ScalarType::Long);
+
+  // Require a tensor accessor for `sobolstate`
+  auto ss_a = wsobolstate.accessor<int64_t, 2>();
+
+  // For every tensor in the list of tensors, the diagonals are made 1
+  // Require the rows of each of the matrices in `ltm`.
+  // Why is this caching these rows separately okay?
+  // A simple calculation show the number of slices to be dimension * MAXBIT * MAXBIT
+  // Instead, by performing `dimension` chunking operations (1 per lower triangular
+  // matrix), we can save some time, since MAXBIT = 30.
+  Tensor eye_maxbit = at::native::eye(MAXBIT, at::ScalarType::Byte);
+  std::vector<std::vector<Tensor>> ltm_rows;
+  for (int64_t d = 0; d < dimension; ++d) {
+    ltm_rows.emplace_back(at::native::chunk(at::native::where(eye_maxbit,
```

Comment:
`at::where` is complaining that the first argument should be of type `Variable` but it is `at::kByte` instead. I am not sure how to fix this.